File: lib/tripit/core/oauth.rb
require 'base64'
require 'cgi'
require 'openssl'
require 'securerandom'
require 'tripit/core/api'

module TripIt
  module Core
    module OAuth
      module Request
        def self.generate_headers(consumer_key:,
                                  consumer_secret:,
                                  nonce:,
                                  timestamp:)
          uri = "https://api.tripit.com/oauth/request_token"
          signature =  TripIt::Core::OAuth.generate_signature(
            uri: uri,
            method: 'GET',
            consumer_key: consumer_key,
            consumer_secret: consumer_secret,
            nonce: nonce,
            timestamp: timestamp)
          return TripIt::Core::OAuth.generate_auth_header(
            realm: uri,
            signature: signature,
            headers:  {
              oauth_consumer_key: consumer_key,
              oauth_nonce: nonce,
              oauth_timestamp: timestamp.to_s,
              oauth_signature_method: 'HMAC-SHA1',
              oauth_version: '1.0',
            })
        end

        def self.get_tokens
          ['TRIPIT_APP_CLIENT_ID','TRIPIT_APP_CLIENT_SECRET'].each {|required|
            raise "Parameter missing: #{required}" if ENV[required].nil?
          }
          timestamp = Time.now.to_i
          auth_header = self.generate_headers(
            consumer_key: ENV['TRIPIT_APP_CLIENT_ID'],
            consumer_secret: ENV['TRIPIT_APP_CLIENT_SECRET'],
            nonce: SecureRandom.hex,
            timestamp:  timestamp)
          url = 'https://api.tripit.com/oauth/request_token'
          response = HTTParty.get(url, {
            headers: { 'Authorization': auth_header }
          })
          if response.code != 200
            raise "Failed to get request token: #{response.body}"
          end
          token_data = {}
          response.body.split('&').each do |attribute|
            key = attribute.split('=').first
            value = attribute.split('=').last
            case key
            when 'oauth_token'
              token_data[:token] = value
            when 'oauth_token_secret'
              token_data[:token_secret] = value
            end
          end
          token_data
        end
      end

      module Access
        def self.generate_headers(consumer_key:,
                                  consumer_secret:,
                                  nonce:,
                                  token:,
                                  token_secret:,
                                  timestamp:)
          uri = "https://api.tripit.com/oauth/access_token"
          signature =  TripIt::Core::OAuth.generate_signature(
            uri: uri,
            method: 'GET',
            consumer_key: consumer_key,
            consumer_secret: consumer_secret,
            token: token,
            token_secret: token_secret,
            nonce: nonce,
            timestamp: timestamp)
          return TripIt::Core::OAuth.generate_auth_header(
            realm: uri,
            signature: signature,
            headers:  {
              oauth_consumer_key: consumer_key,
              oauth_token: token,
              oauth_nonce: nonce,
              oauth_timestamp: timestamp.to_s,
              oauth_signature_method: 'HMAC-SHA1',
              oauth_version: '1.0',
            })
        end

        def self.get_tokens(request_token:, request_token_secret:)
          ['TRIPIT_APP_CLIENT_ID','TRIPIT_APP_CLIENT_SECRET'].each {|required|
            raise "Parameter missing: #{required}" if ENV[required].nil?
          }
          timestamp = Time.now.to_i
          auth_header = self.generate_headers(
            consumer_key: ENV['TRIPIT_APP_CLIENT_ID'],
            consumer_secret: ENV['TRIPIT_APP_CLIENT_SECRET'],
            nonce: SecureRandom.hex,
            token: request_token,
            token_secret: request_token_secret,
            timestamp:  timestamp)
          url = 'https://api.tripit.com/oauth/access_token'
          response = HTTParty.get(url, {
            headers: { 'Authorization': auth_header }
          })
          if response.code != 200
            raise "Failed to get request token: #{response.body}"
          end
          token_data = {}
          response.body.split('&').each do |attribute|
            key = attribute.split('=').first
            value = attribute.split('=').last
            case key
            when 'oauth_token'
              token_data[:token] = value
            when 'oauth_token_secret'
              token_data[:token_secret] = value
            end
          end
          token_data
        end
      end

      # Retrieve an OAuth token from a given code and client ID/secret.
      def self.access(client_id:, client_secret:, code:, redirect_uri:)
        params = {
          client_id: client_id,
          client_secret: client_secret,
          code: code,
          redirect_uri: redirect_uri
        }
        TripIt::Core::API.post_to(endpoint: 'oauth.access',
                                  content_type: 'application/x-www-formencoded',
                                  params: params)
      end

      def self.token_expired?(token:)
        response = TripIt::Core::API.get_from(endpoint: 'auth.test',
                                              params: { token: token })
        json = JSON.parse(response.body, symbolize_names: true)
        json[:ok] == false and json[:error] == 'invalid_auth'
      end

      def self.generate_auth_header(realm:, headers:, signature:)
        [
          "OAuth realm=\"#{realm}\"",
          headers.sort.to_h.map{|key,value| "#{key}=\"#{value}\""}.join(','),
          "oauth_signature=\"#{CGI.escape(signature)}\""
        ].join(',')
      end

      def self.generate_signature(method:, uri:, consumer_key:,
                                  nonce:, timestamp:, token: nil,
                                  token_secret: nil, consumer_secret:)
        params = {
          oauth_consumer_key: consumer_key,
          oauth_nonce: nonce,
          oauth_signature_method: 'HMAC-SHA1',
          oauth_timestamp: timestamp,
          oauth_version: '1.0'
        }
        if !token.nil?
          params[:oauth_token] = token
        end
        params = params.sort.to_h
        encrypt_key = "#{consumer_secret}&" # This isn't in the TripIt documentation!
        if !token_secret.nil?
          encrypt_key += token_secret
        end
        require 'pp'
        params_serialized = params.map{|key,value| "#{key}=#{value}"}.join('&')
        sig_base_string = "#{method}&#{CGI.escape uri}&#{CGI.escape params_serialized}"
        raw_signature = OpenSSL::HMAC.digest('SHA1', encrypt_key, sig_base_string)
        encoded_signature = Base64.encode64(raw_signature).strip
        return encoded_signature
      end
    end
  end
end

File: tripit/core/v1/oauth.py
"""
Functions for authenticating calls to TripIt via OAuth v1.
"""
from datetime import datetime
import urllib.parse
from hashlib import sha1
import base64
import os
import hmac
import logging
import secrets
import requests
from tripit.environment import EnvironmentCheck
from tripit.logging import logger
from tripit.helpers import sort_dict


def get_missing_client_id_and_secret_env_vars():
    """ tfw the function name is the documentation """
    env_check = EnvironmentCheck(["TRIPIT_APP_CLIENT_SECRET", "TRIPIT_APP_CLIENT_ID"])
    return env_check.missing_vars


def fetch_token(token=None, token_secret=None):
    """
    Get a new request token from the TripIt API.
    """
    env_vars = get_missing_client_id_and_secret_env_vars()
    if env_vars:
        raise RuntimeError(f"Please define these environment variables: {env_vars}")

    client_id = os.environ.get("TRIPIT_APP_CLIENT_ID")
    client_secret = os.environ.get("TRIPIT_APP_CLIENT_SECRET")
    timestamp = int(datetime.now().timestamp())
    nonce = secrets.token_hex()
    """ If we are trying to request tokens and already have a token
        secret, then that means we already went through the first step
        of the OAuth process and are now trying to get access tokens. """
    if token_secret is not None:
        request_uri = "https://api.tripit.com/oauth/access_token"
    else:
        request_uri = "https://api.tripit.com/oauth/request_token"
    logger.debug("Token: %s, token_secret: %s, URI: %s", token, token_secret, request_uri)
    common_arguments = {
        "uri": request_uri,
        "consumer_key": client_id,
        "nonce": nonce,
        "timestamp": timestamp,
    }
    access_token_arguments = {}
    if token_secret is not None:
        access_token_arguments["token"] = token
        access_token_arguments["token_secret"] = token_secret

    oauth_sig = generate_signature(
        method="GET", consumer_secret=client_secret, **common_arguments, **access_token_arguments
    )
    auth_header = generate_sha1_auth_header(
        signature=oauth_sig, **common_arguments, **access_token_arguments
    )
    logger.debug("Auth header: %s", auth_header)
    response = requests.get(request_uri, headers={"Authorization": auth_header})
    if response.status_code != 200:
        logging.error("Failed to get token data: %s)", response.text)
        return None
    token_data = {}
    for token_part in response.text.split("&"):
        key, value = token_part.split("=")
        token_data[key.replace("oauth_", "")] = value
    return token_data


def request_request_token():
    """ Request a request token.
    This is here temporarily while I refactor request_token(). """
    return fetch_token()


def request_access_token(req_token, request_token_secret):
    """ Fetch an access token after fetching a request token. """
    return fetch_token(req_token, request_token_secret)


# pylint: disable=too-many-arguments
def generate_authenticated_headers_for_request(
    method, uri, consumer_key, consumer_secret, token, token_secret
):
    """ Generates heades for authenticated API calls. """
    nonce = secrets.token_hex()
    timestamp = datetime.now().timestamp()
    signature = generate_signature(
        method, uri, consumer_key, consumer_secret, nonce, timestamp, token, token_secret
    )
    return generate_sha1_auth_header(uri, signature, consumer_key, nonce, timestamp, token)


# pylint: disable=too-many-arguments
def generate_sha1_auth_header(uri, signature, consumer_key, nonce, timestamp, token=None, **kwargs):
    """
    Generates an OAuth v1 authencation header for HTTP requests to endpoints
    requiring OAuth v1 authentication.
    """
    headers = {
        "oauth_consumer_key": consumer_key,
        "oauth_nonce": nonce,
        "oauth_signature_method": "HMAC-SHA1",
        "oauth_timestamp": int(timestamp),
        "oauth_token": token or "",
        "oauth_version": "1.0",
    }
    if token is None:
        headers.pop("oauth_token")
    encoded_sig = urllib.parse.quote_plus(signature)
    auth_header_parts = [
        f'OAuth realm="{uri}"',
        ",".join([f'{k}="{v}"' for k, v in sort_dict(headers).items()]),
        f'oauth_signature="{encoded_sig}"',
    ]
    return ",".join(auth_header_parts)


# pylint: disable=too-many-arguments
def generate_signature(
    method, uri, consumer_key, consumer_secret, nonce, timestamp, token=None, token_secret=None
):
    """
    Generates an OAuth v1 signature. These are used to form authentication headers.

    Unfortunately, because we really require these many arguments while working
    with OAuth v1, we need to tell pylint to disable the, usually correct,
    "too-many-arguments" error. Otherwise, we'll need to resort to using
    **kwargs, which is too unsafe for my liking.
    """
    params = {
        "oauth_consumer_key": consumer_key,
        "oauth_nonce": nonce,
        "oauth_signature_method": "HMAC-SHA1",
        "oauth_timestamp": int(timestamp),
        "oauth_version": "1.0",
    }
    if token:
        params["oauth_token"] = token

    encrypt_key = "&".join([consumer_secret, (token_secret if token_secret is not None else "")])
    param_parts = "&".join([f"{key}={params[key]}" for key, value in sort_dict(params).items()])
    base_string_for_signature = "&".join(
        [method, urllib.parse.quote_plus(uri), urllib.parse.quote_plus(param_parts)]
    )
    signature = hmac.new(bytes(encrypt_key, "utf8"), bytes(base_string_for_signature, "utf8"), sha1)

    # Trying to figure out why access tokens have bad sigs
    for param in params.keys():
        logger.debug("%s: %s", param, params[param])
    logger.debug("Encryption key, if any: %s", encrypt_key)
    logger.debug("Signature base: %s", base_string_for_signature)
    logger.debug("Signature: %s", base64.b64encode(signature.digest()))
    logger.debug("Method: %s, URI: %s", method, uri)
    return base64.b64encode(signature.digest())
